TestClientServer_FpRegisters fails for XMM12 registers

1. What version of Delve are you using (`dlv version`)?
   1.23.1
2. What version of Go are you using? (`go version`)?
   1.23.1
3. What operating system and processor architecture are you using?
   Debian sid. amd64. 4th generation EPYC processors from AMD
4. What did you do?
   Package delve for Debian and run the tests during build.
5. What did you expect to see?
   No error.
6. What did you see instead?
```
     integration2_test.go:1806: register XMM12 expected to contain "…[ZMM12hl] 0x3ff66666666666663ff4cccccccccccd" got "0x3ff66666666666663ff4cccccccccccd\tv2_int={ 3ff4cccccccccccd 3ff6666666666666 }\tv4_int={ cccccccd 3ff4cccc 66666666 3ff66666 }\tv8_int={ cccd cccc cccc 3ff4 6666 6666 6666 3ff6 }\tv16_int={ cd cc cc cc cc cc f4 3f 66 66 66 66 66 66 f6 3f }\tv2_float={ 1.3 1.4 }\tv4_float={ -1.07374184e+08 1.9124999 2.720083e+23 1.925 }\n\t[YMM12h] 0x3ff66666666666663ff4cccccccccccd\tv2_int={ 3ff4cccccccccccd 3ff6666666666666 }\tv4_int={ cccccccd 3ff4cccc 66666666 3ff66666 }\tv8_int={ cccd cccc cccc 3ff4 6666 6666 6666 3ff6 }\tv16_int={ cd cc cc cc cc cc f4 3f 66 66 66 66 66 66 f6 3f }\tv2_float={ 1.3 1.4 }\tv4_float={ -1.07374184e+08 1.9124999 2.720083e+23 1.925 }\n\t[ZMM12hl] 0x00000000000000000000000000000000\tv2_int={ 0000000000000000 0000000000000000 }\tv4_int={ 00000000 00000000 00000000 00000000 }\tv8_int={ 0000 0000 0000 0000 0000 0000 0000 0000 }\tv16_int={ 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 }\tv2_float={ 0 0 }\tv4_float={ 0 0 0 0 }\n\t[ZMM12hh] 0x00000000000000000000000000000000\tv2_int={ 0000000000000000 0000000000000000 }\tv4_int={ 00000000 00000000 00000000 00000000 }\tv8_int={ 0000 0000 0000 0000 0000 0000 0000 0000 }\tv16_int={ 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 }\tv2_float={ 0 0 }\tv4_float={ 0 0 0 0 }"

```

It's reported at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1082973
Full log https://people.debian.org/~sanvila/build-logs/202409/delve_1.23.1-1_amd64-20240929T101703.749Z

Unfortunately I don't have CPU with AVX512, so I can't confirm the failure myself.

Reproducible on Zen4
```
=== RUN   TestClientServer_FpRegisters
    support.go:246: enabling recording for TestClientServer_FpRegisters
    integration2_test.go:1765: state after continue: &api.DebuggerState{Pid:4436, TargetCommandLine:"/tmp/fputest.fe32f2ed", Running:false, Recording:false, CoreDumping:false, CurrentThread:(*api.Thread)(0xc0001dacb0), SelectedGoroutine:(*api.Goroutine)(0xc000135860), Threads:[]*api.Thread{(*api.Thread)(0xc0001dad90), (*api.Thread)(0xc0001dae70), (*api.Thread)(0xc0001daee0), (*api.Thread)(0xc0001daf50), (*api.Thread)(0xc0001dafc0)}, NextInProgress:false, WatchOutOfScope:[]*api.Breakpoint(nil), Exited:false, ExitStatus:0, When:"", Err:error(nil)}
    integration2_test.go:1774: avx2 variable: &api.Variable{Name:"avx2", Addr:0xc0000a4ef2, OnlyAddr:false, Type:"bool", RealType:"bool", Flags:0x0, Kind:0x1, Value:"true", Len:0, Cap:0, Children:[]api.Variable{}, Base:0x0, Unreadable:"", LocationExpr:"[block] DW_OP_fbreg -0x5e ", DeclLine:22}
    integration2_test.go:1774: avx512 variable: &api.Variable{Name:"avx512", Addr:0xc0000a4ef1, OnlyAddr:false, Type:"bool", RealType:"bool", Flags:0x0, Kind:0x1, Value:"true", Len:0, Cap:0, Children:[]api.Variable{}, Base:0x0, Unreadable:"", LocationExpr:"[block] DW_OP_fbreg -0x5f ", DeclLine:23}
    integration2_test.go:1788: state after continue: &api.DebuggerState{Pid:4436, TargetCommandLine:"/tmp/fputest.fe32f2ed", Running:false, Recording:false, CoreDumping:false, CurrentThread:(*api.Thread)(0xc0001db730), SelectedGoroutine:(*api.Goroutine)(0xc000135ba0), Threads:[]*api.Thread{(*api.Thread)(0xc0001db810), (*api.Thread)(0xc0001db880), (*api.Thread)(0xc0001db8f0), (*api.Thread)(0xc0001db9d0), (*api.Thread)(0xc0001dba40)}, NextInProgress:false, WatchOutOfScope:[]*api.Breakpoint(nil), Exited:false, ExitStatus:0, When:"", Err:error(nil)}
    integration2_test.go:1806: register XMM12 expected to contain "…[ZMM12hl] 0x3ff66666666666663ff4cccccccccccd" got "0x3ff66666666666663ff4cccccccccccd\tv2_int={ 3ff4cccccccccccd 3ff6666666666666 }\tv4_int={ cccccccd 3ff4cccc 66666666 3ff66666 }\tv8_int={ cccd cccc cccc 3ff4 6666 6666 6666 3ff6 }\tv16_int={ cd cc cc cc cc cc f4 3f 66 66 66 66 66 66 f6 3f }\tv2_float={ 1.3 1.4 }\tv4_float={ -1.07374184e+08 1.9124999 2.720083e+23 1.925 }\n\t[YMM12h] 0x3ff66666666666663ff4cccccccccccd\tv2_int={ 3ff4cccccccccccd 3ff6666666666666 }\tv4_int={ cccccccd 3ff4cccc 66666666 3ff66666 }\tv8_int={ cccd cccc cccc 3ff4 6666 6666 6666 3ff6 }\tv16_int={ cd cc cc cc cc cc f4 3f 66 66 66 66 66 66 f6 3f }\tv2_float={ 1.3 1.4 }\tv4_float={ -1.07374184e+08 1.9124999 2.720083e+23 1.925 }\n\t[ZMM12hl] 0x00000000000000000000000000000000\tv2_int={ 0000000000000000 0000000000000000 }\tv4_int={ 00000000 00000000 00000000 00000000 }\tv8_int={ 0000 0000 0000 0000 0000 0000 0000 0000 }\tv16_int={ 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 }\tv2_float={ 0 0 }\tv4_float={ 0 0 0 0 }\n\t[ZMM12hh] 0x00000000000000000000000000000000\tv2_int={ 0000000000000000 0000000000000000 }\tv4_int={ 00000000 00000000 00000000 00000000 }\tv8_int={ 0000 0000 0000 0000 0000 0000 0000 0000 }\tv16_int={ 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 }\tv2_float={ 0 0 }\tv4_float={ 0 0 0 0 }"
--- FAIL: TestClientServer_FpRegisters (0.25s)
FAIL
exit status 1
FAIL    github.com/go-delve/delve/service/test  0.259s
```

--- a/pkg/proc/amd64util/xsave.go
+++ b/pkg/proc/amd64util/xsave.go
@@ -79,7 +79,6 @@ const (
 	_XSAVE_HEADER_LEN              = 64
 	_XSAVE_EXTENDED_REGION_START   = 576
 	_XSAVE_SSE_REGION_LEN          = 416
-	_XSAVE_AVX512_ZMM_REGION_START = 1152
 )
 
 // AMD64XstateRead reads a byte array containing an XSAVE area into regset.
@@ -87,7 +86,8 @@ const (
 // contents of the legacy region of the XSAVE area.
 // See Section 13.1 (and following) of Intel® 64 and IA-32 Architectures
 // Software Developer’s Manual, Volume 1: Basic Architecture.
-func AMD64XstateRead(xstateargs []byte, readLegacy bool, regset *AMD64Xstate) error {
+// If xstateZMMHi256Offset is zero, it will be guessed.
+func AMD64XstateRead(xstateargs []byte, readLegacy bool, regset *AMD64Xstate, xstateZMMHi256Offset int) error {
 	if _XSAVE_HEADER_START+_XSAVE_HEADER_LEN >= len(xstateargs) {
 		return nil
 	}
@@ -120,7 +120,19 @@ func AMD64XstateRead(xstateargs []byte, readLegacy bool, regset *AMD64Xstate) er
 		return nil
 	}
 
-	avx512state := xstateargs[_XSAVE_AVX512_ZMM_REGION_START:]
+	if xstateZMMHi256Offset == 0 {
+		// Guess ZMM_Hi256 component offset
+		// ref: https://github.com/bminor/binutils-gdb/blob/df89bdf0baf106c3b0a9fae53e4e48607a7f3f87/gdb/i387-tdep.c#L916
+		if xstate_bv&(1<<9) != 0 && len(xstateargs) == 2440 {
+			// AMD CPUs supporting PKRU
+			xstateZMMHi256Offset = 896
+		} else {
+			// Intel CPUs supporting AVX512
+			xstateZMMHi256Offset = 1152
+		}
+	}
+
+	avx512state := xstateargs[xstateZMMHi256Offset:]
 	regset.Avx512State = true
 	copy(regset.ZmmSpace[:], avx512state[:len(regset.ZmmSpace)])
 
@@ -180,7 +192,7 @@ func (xstate *AMD64Xstate) SetXmmRegister(n int, value []byte) error {
 	// Copy bytes [32, 64) to Xsave area
 
 	zmmval := rest
-	zmmpos := _XSAVE_AVX512_ZMM_REGION_START + (n * 32)
+	zmmpos := AMD64XstateZMMHi256Offset() + (n * 32)
 	if zmmpos >= len(xstate.Xsave) {
 		return fmt.Errorf("could not set XMM%d: bytes 32..%d not in XSAVE area", n, 32+len(zmmval))
 	}

--- a/pkg/proc/amd64util/xsave_amd64.go
+++ b/pkg/proc/amd64util/xsave_amd64.go
@@ -28,3 +28,28 @@ func AMD64XstateMaxSize() int {
 	})
 	return xstateMaxSize
 }
+
+var xstateZMMHi256Offset int
+var loadXstateZMMHi256OffsetOnce sync.Once
+
+// AMD64XstateZMMHi256Offset probes ZMM_Hi256 offset of the current CPU.  Beware
+// that core dumps may be generated from a different CPU.
+func AMD64XstateZMMHi256Offset() int {
+	loadXstateZMMHi256OffsetOnce.Do(func() {
+		// See Intel 64 and IA-32 Architecture Software Developer's Manual, Vol. 1
+		// chapter 13.2 and Vol. 2A CPUID instruction for a description of all the
+		// magic constants.
+
+		_, _, cx, _ := cpuid(0x01, 0x00)
+
+		if cx&(1<<26) == 0 { // Vol. 2A, Table 3-10, XSAVE enabled bit check
+			// XSAVE not supported by this processor
+			xstateZMMHi256Offset = 0
+			return
+		}
+
+		_, bx, _, _ := cpuid(0x0d, 0x06) // ZMM_Hi256 is component #6
+		xstateZMMHi256Offset = int(bx)
+	})
+	return xstateZMMHi256Offset
+}

--- a/pkg/proc/amd64util/xsave_other.go
+++ b/pkg/proc/amd64util/xsave_other.go
@@ -5,3 +5,8 @@ package amd64util
 func AMD64XstateMaxSize() int {
 	return _XSTATE_MAX_KNOWN_SIZE
 }
+
+func AMD64XstateZMMHi256Offset() int {
+	// AVX-512 not supported
+	return 0
+}

--- a/pkg/proc/core/linux_core.go
+++ b/pkg/proc/core/linux_core.go
@@ -351,7 +351,7 @@ func readNote(r io.ReadSeeker, machineType elf.Machine) (*note, error) {
 	case _NT_X86_XSTATE:
 		if machineType == _EM_X86_64 {
 			var fpregs amd64util.AMD64Xstate
-			if err := amd64util.AMD64XstateRead(desc, true, &fpregs); err != nil {
+			if err := amd64util.AMD64XstateRead(desc, true, &fpregs, 0); err != nil {
 				return nil, err
 			}
 			note.Desc = &fpregs

--- a/pkg/proc/native/ptrace_freebsd_amd64.go
+++ b/pkg/proc/native/ptrace_freebsd_amd64.go
@@ -73,7 +73,7 @@ func ptraceGetRegset(id int) (*amd64util.AMD64Xstate, error) {
 	if err != nil {
 		return nil, err
 	}
-	err = amd64util.AMD64XstateRead(regset.Xsave, false, &regset)
+	err = amd64util.AMD64XstateRead(regset.Xsave, false, &regset, amd64util.AMD64XstateZMMHi256Offset())
 	return &regset, err
 }
 

--- a/pkg/proc/native/ptrace_linux_386.go
+++ b/pkg/proc/native/ptrace_linux_386.go
@@ -38,7 +38,7 @@ func ptraceGetRegset(tid int) (regset amd64util.AMD64Xstate, err error) {
 	}
 
 	regset.Xsave = xstateargs[:iov.Len]
-	err = amd64util.AMD64XstateRead(regset.Xsave, false, &regset)
+	err = amd64util.AMD64XstateRead(regset.Xsave, false, &regset, amd64util.AMD64XstateZMMHi256Offset())
 	return
 }
 

--- a/pkg/proc/native/ptrace_linux_amd64.go
+++ b/pkg/proc/native/ptrace_linux_amd64.go
@@ -37,6 +37,6 @@ func ptraceGetRegset(tid int) (regset amd64util.AMD64Xstate, err error) {
 	}
 
 	regset.Xsave = xstateargs[:iov.Len]
-	err = amd64util.AMD64XstateRead(regset.Xsave, false, &regset)
+	err = amd64util.AMD64XstateRead(regset.Xsave, false, &regset, amd64util.AMD64XstateZMMHi256Offset())
 	return
 }

